presence_channel.go

```diff
@@ -0,0 +1,249 @@
+// subscribed.
+//
+// Note: Bind() does not fire pusher:member_added/removed, use
+// BindMemberAdded/Removed() instead.
+type PresenceChannel interface {
+	Channel
+	// BindMemberAdded returns a channel that receives a Member value when a user
+	// joins the channel.
+	BindMemberAdded() chan Member
+	// BindMemberRemoved returns a channel that receives a user ID when a user
+	// leaves the channel.
+	BindMemberRemoved() chan string
+	// Members returns the member info of each user currently subscribed to the
+	// channel.
+	Members() map[string]Member
+	// Member returns a representation of the channel member with the given ID.
+	// If the member isn't in the channel, the returned bool is `false`.
+	Member(id string) (bool, Member)
+	// Me returns the member for the current user.
+	//
+	// Possible errors:
+	//  - not subscribed - subscription must succeed before calling Me()
+	//  - invalid channel data - pusher server provided invalid JSON
+	//  - missing member for current user - pusher server violated protocol
+	Me() (*Member, error)
+	// MemberCount returns the number of users connected to the channel.
+	MemberCount() int
+}
+
+// presenceChannel implements the internalChannel and PresenceChannel interfaces
+type presenceChannel struct {
+	*privateChannel
+
+	memberAdded   chan Member
+	memberRemoved chan string
+	members       map[string]Member
+}
+
+func newPresenceChannel(baseChannel *channel) *presenceChannel {
+	privateChannel := &privateChannel{channel: baseChannel}
+	return &presenceChannel{
+		privateChannel: privateChannel,
+		memberAdded:    make(chan Member),
+		memberRemoved:  make(chan string),
+		members:        map[string]Member{},
+	}
+}
+
+// presenceChannelSubscriptionData is sent from Pusher to the client in the
+// subscription_succeeded event.
+// https://pusher.com/docs/channels/library_auth_reference/pusher-websockets-protocol/#pusher_internalsubscription_succeeded-pusher-channels-greater-client
+type presenceChannelSubscriptionData struct {
+	Presence struct {
+		IDs   []string                   `json:"ids"`
+		Hash  map[string]json.RawMessage `json:"hash"`
+		Count int                        `json:"count"`
+	} `json:"presence"`
+}
+
+type presenceChannelData struct {
+	UserID string `json:"user_id"`
+}
+
+type presenceChannelMemberAddedData struct {
+	UserID   string          `json:"user_id"`
+	UserInfo json.RawMessage `json:"user_info"`
+}
+
+type presenceChannelMemberRemovedData struct {
+	UserID string `json:"user_id"`
+}
+
+func (c *presenceChannel) handleEvent(event string, data json.RawMessage) {
+	if event == pusherInternalMemberAdded {
+		var member presenceChannelMemberAddedData
+		err := UnmarshalDataString(data, &member)
+		if err != nil {
+			c.client.sendError(err)
+		} else {
+			c.mutex.Lock()
+			c.members[member.UserID] = Member{
+				ID:   member.UserID,
+				Info: member.UserInfo,
+			}
+			c.mutex.Unlock()
+		}
+
+		select {
```

Comment:
> I think the reason for the select/default approach was to avoid blocking if there are no goroutines processing the messages, right?

Yup, exactly!

> If that's the case, I think we may need to build that into the API. I.e. we need to be able to tell when a user has binded a goroutine, and only send messages then. We would also need a way for the client to unbind.

Now I'm not totally sure about this... how could we avoid the situation you mentioned above -- the consumer receives a message but doesn't listen again before the next message?

Igor noted that the current implementation wouldn't work across multiple goroutines (since messages would only be delivered to one goroutines -- whoops). I'm going to refactor BindMemberAdded/Removed to work like `Bind` where a channel is created on each invocation.

Reply:
Correct. `select/default` will only send the message if there's currently a goroutine waiting to receive the message.

E.g. in the following example:

```go
for x := range myChannel {
  // do task that takes 500ms
}
```

While that goroutine is doing the task that takes 500ms, `select/default` style sends will be dropped. Another one will only be sent successfully once its done with the loop body.

Given the same example, if the producer uses a normal send, the producer will block until the receiver is ready to receive a new message.

Reply:
I guess one way is using a goroutine as a buffer like this:

```
for _, c := range chansMap {
  go func() {
    c <- message
  }()
}
```

Which is how `Bind` works

Reply:
As far as supporting multiple goroutines goes, yep creating a new channel for each `Bind*` call seems like the way to go. However if we move toward blocking sends, we should also add a way to remove those channels because otherwise it's possible to deadlock if one of those consumer goroutines stops processing messages.

Reply:
yep the buffer goroutine approach would prevent blocking the `handleEvent` goroutine. However, I think we'd still need a way to unbind the consumers and stop those goroutines, otherwise we'll have a goroutine leak instead of a deadlock – which I guess is better, but still not great 😁

Reply:
Implemented a solution that uses context to unblock goroutines when `Unbind`/`UnbindMemberAdded/Removed` is used for a channel.